executor/read/step_executor.rs

```diff
@@ -0,0 +1,251 @@
+        executable_stages.len() - 1,
+        tmp__recursion_validation,
+    )?;
+
+    // TODO: Add table writing step.
+    match &executable_function.returns {
+        ExecutableReturn::Stream(positions) => {
+            steps.push(StepExecutors::ReshapeForReturn(positions.clone()));
+        }
+        _ => todo!(),
+    }
+    Ok(steps)
+}
+
+pub(super) fn create_executors_for_pipeline_stages(
+    snapshot: &Arc<impl ReadableSnapshot + 'static>,
+    thing_manager: &Arc<ThingManager>,
+    function_registry: &ExecutableFunctionRegistry,
+    executable_stages: &Vec<ExecutableStage>,
+    at_index: usize,
+    tmp__recursion_validation: &mut HashSet<FunctionID>,
+) -> Result<Vec<StepExecutors>, ConceptReadError> {
+    let mut previous_stage_steps = if at_index > 0 {
+        create_executors_for_pipeline_stages(
+            snapshot,
+            thing_manager,
+            function_registry,
+            executable_stages,
+            at_index - 1,
+            tmp__recursion_validation,
+        )?
+    } else {
+        vec![]
+    };
+
+    match &executable_stages[at_index] {
+        ExecutableStage::Match(match_executable) => {
+            let mut match_stages = create_executors_for_match(
+                snapshot,
+                thing_manager,
+                function_registry,
+                match_executable,
+                tmp__recursion_validation,
+            )?;
+            previous_stage_steps.append(&mut match_stages);
+            Ok(previous_stage_steps)
+        }
+        ExecutableStage::Select(_) => todo!(),
+        ExecutableStage::Offset(offset_executable) => {
+            let inner = NestedPatternBranch::new(PatternExecutor::new(previous_stage_steps));
+            let mapper = OffsetMapper::new(offset_executable.offset);
+            let step = NestedPatternExecutor::Offset([inner], mapper);
+            Ok(vec![StepExecutors::Branch(step)])
+        }
+        ExecutableStage::Limit(limit_executable) => {
+            let inner = NestedPatternBranch::new(PatternExecutor::new(previous_stage_steps));
+            let mapper = LimitMapper::new(limit_executable.limit);
+            let step = NestedPatternExecutor::Limit([inner], mapper);
+            Ok(vec![StepExecutors::Branch(step)])
+        }
+        ExecutableStage::Require(_) => todo!(),
+        ExecutableStage::Sort(sort_executable) => {
+            let step = CollectingStageExecutor::new_sort(PatternExecutor::new(previous_stage_steps), sort_executable);
+            Ok(vec![StepExecutors::CollectingStage(step)])
+        }
+        ExecutableStage::Reduce(reduce_executable) => {
+            let step = CollectingStageExecutor::new_reduce(
+                PatternExecutor::new(previous_stage_steps),
+                reduce_executable.clone(),
+            );
+            Ok(vec![StepExecutors::CollectingStage(step)])
+        }
+        ExecutableStage::Insert(_) | ExecutableStage::Delete(_) => todo!("Or unreachable?"),
```

Comment:
I think having a `todo` with a question `unreachable?` makes no sense :D Either it is a TODO or it is unreachable - I think these kinds of questions marks should be resolved before merge

Reply:
There's another one in plan.rs

Reply:
I suppose the question is: Can you use functions in a write pipeline?
In which case this will likely have to be split out into two code paths - One for the entry pattern and one for functions.